core/src/main/java/io/questdb/griffin/engine/table/parquet/PartitionUpdater.java

```diff
@@ -0,0 +1,206 @@
+
+package io.questdb.griffin.engine.table.parquet;
+
+import io.questdb.cairo.CairoException;
+import io.questdb.cairo.TableUtils;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.*;
+import io.questdb.std.str.LPSZ;
+
+public class PartitionUpdater implements QuietCloseable {
+    private static final Log LOG = LogFactory.getLog(PartitionUpdater.class);
+    private long fd;
+    private long ptr;
+    private final FilesFacade ff;
+
+    public PartitionUpdater(FilesFacade ff) {
+        this.ff = ff;
+    }
+
+    // Currently unused; will be used for optimisation in the future.
+    public void appendRowGroup(PartitionDescriptor descriptor) {
+        final int columnCount = descriptor.getColumnCount();
+        final long rowCount = descriptor.getPartitionRowCount();
+        try {
+            appendRowGroup(
+                    ptr,
+                    columnCount,
+                    descriptor.getColumnNamesPtr(),
+                    descriptor.getColumnNamesSize(),
+                    descriptor.getColumnNameLengthsPtr(),
+                    descriptor.getColumnTypesPtr(),
+                    descriptor.getColumnIdsPtr(),
+                    descriptor.getColumnTopsPtr(),
+                    descriptor.getColumnAddressesPtr(),
+                    descriptor.getColumnSizesPtr(),
+                    descriptor.getColumnSecondaryAddressesPtr(),
+                    descriptor.getColumnSecondarySizesPtr(),
+                    descriptor.getSymbolOffsetsAddressesPtr(),
+                    descriptor.getSymbolOffsetsSizesPtr(),
+                    rowCount
+            );
+        } catch (Throwable th) {
+            throw CairoException.critical(0).put("Could not append new rowGroup: [table=").put(descriptor.getTableName())
+                    .put(", exception=").put(th.getClass().getSimpleName())
+                    .put(", msg=").put(th.getMessage())
+                    .put(']');
+        } finally {
+            descriptor.clear();
+        }
+    }
+
+    @Override
+    public void close() {
+        destroy();
+        fd = -1;
+    }
+
+    public void of(
+            @Transient LPSZ srcPath,
+            long fileOpenOpts,
+            long fileSize,
+            int timestampIndex,
+            long compressionCodec,
+            boolean statisticsEnabled,
+            long rowGroupSize,
+            long dataPageSize
+    ) {
+        destroy();
+        this.fd = TableUtils.openRW(ff, srcPath, LOG, fileOpenOpts);
+        try {
+            ptr = create(
+                    Files.detach(fd),
+                    fileSize,
+                    timestampIndex,
+                    compressionCodec,
+                    statisticsEnabled,
+                    rowGroupSize,
+                    dataPageSize
+            );
+        } catch (Throwable th) {
+            throw CairoException.nonCritical().put("could not open parquet file for update: [path=").put(srcPath)
```

Comment:
If ptr is not initialized, fd is never closed. Either close fd here or in the close when ptr is 0.

Reply:
PartitionUpdater (Java side) is not the owner of fd, so it should not store it. Fixed.